src/sink/nats.rs

```diff
@@ -0,0 +1,298 @@
+                    }
+                    // prepare message headers
+                    let mut key_val: Vec<(&str, &str)> = Vec::new();
+                    if let Some(headers) = headers {
+                        if let Some(headers_obj) = headers.as_object() {
+                            for (key, val) in headers_obj.iter() {
+                                if let Some(array) = val.as_array() {
+                                    for ele in array.iter() {
+                                        if let Some(ele) = ele.as_str() {
+                                            key_val.push((key, ele));
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    } else {
+                        for (key, val) in &self.config.headers {
+                            for ele in val.iter() {
+                                key_val.push((key.as_str(), ele.as_str()));
+                            }
+                        }
+                    }
+                    let key_val_len = key_val.len();
+                    let nats_headers = Headers::from_iter(key_val);
+                    let message_headers = if key_val_len > 0 {
+                        Some(&nats_headers)
+                    } else {
+                        None
+                    };
+
+                    // this can silently fail if the server doesnt support message headers.
+                    // check https://github.com/nats-io/nats.rs/issues/160
+                    let publish_result = connection
+                        .publish_with_reply_or_headers(
+                            self.config.subject.as_str(),
+                            message_reply,
+                            message_headers,
+                            payload,
+                        )
+                        .await;
+                    match publish_result {
+                        Ok(()) => {
+                            self.reply_channel
+                                .send(sink::Reply::Insight(insight.clone()))
+                                .await?;
+                        }
+                        Err(e) => {
+                            error!("[Sink::{}] failed to send message: {}", &self.sink_url, e);
+                            let error_insight = Event {
+                                cb: CBAction::Fail,
+                                ..insight.clone()
+                            };
+                            self.reply_channel
+                                .send(sink::Reply::Response(ERR, error_insight))
+                                .await?;
+                        }
+                    }
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    async fn on_signal(&mut self, signal: Event) -> ResultVec {
+        if self.connection.is_none() {
```

Comment:
How could we get here?
After what I see, we do fail `init` if we can't connect, so we will only get active if we established a connection, and there is no place where we set the connection to `None`.

Should a background task call `self.connection.flush()` in a regular manner and set the connection to `None` if this fails? Should we try and reconnect when a publish request fails?

Anyways if you send those restores/triggers, you need to put an op_meta into the insight event. Please see the Ws sink and how it handles it with the `merged_meta` field.

Reply:
Looking at the current implementation, it looks like `on_signal()` will be called every 100 ms, so instead of checking `connection` is `None`, we could  call `connection.flush()`, try to restore the connection and set it to `None` if connecting fails? Regarding, reconnecting if a publish request fails, since we are going to be running this very frequently, is that really required? Please let me know :D
I'll take a look at the websocket sink, to get a better idea on how to handle `op_meta`.

Reply:
Calling flush here is not a good idea, as this will `await` until it was able to do the `ping - pong` roundtrip, i think this can take quite long and in this time the sink cannot progress otherwise, which is not what we want. Also, i don't know if it is possible to share a connection across tasks, so a background task might not be feasible.

Reply:
I think we can do this in a background task. If you take a look at the `nats.rs` source code, a [`Connection` seems to be an `Arc` wrapping another struct](https://github.com/nats-io/nats.rs/blob/master/src/lib.rs#L307), which should be safe to pass across tasks, if I am not mistaken.

Reply:
yes, in this case this should be safe. I only hope this won't block publishing events in some way.

Reply:
I was playing around with this, and I think it makes more sense to recover a dead connection when we fail to publish a message rather than call `connection.flush()` in a background task in `on_signal()` (similar to how we handle this in kafka). Do let me know your thoughts about this.

Reply:
The simpler the better! :+1: And it will work greatly if you can somehow distinguish between connection errors and non-connection errors.